### ISSUE TITLE
question about the finetune data

### ISSUE THREAD

**Issue by yanyanxixi** [jik876/hifi-gan#43]
in readme ,you said when finetuned should "Generate mel-spectrograms in numpy format using Tacotron2 with teacher-forcing"
but the generated mel-spectrograms 's number is not equal the oral mel，in meldataset.py.  got the audio like this
if audio.size(1) >= self.segment_size:
                    mel_start = random.randint(0, mel.size(2) - frames_per_seg - 1)
                    mel = mel[:, :, mel_start:mel_start + frames_per_seg]
                    audio = audio[:, mel_start * self.hop_size:(mel_start + frames_per_seg) * self.hop_size]

in this way, if  we use the oral audio but use the generated mel,  the traindata will be wrong, right?

**Comment by yanyanxixi** [comment]
mybe  use teacher-forcing can get the same number of oral wav

**Comment by leminhnguyen** [comment]
@yanyanxixi  I think I have the same question likes you. You have used teacher-forcing in tacotron training or hifigan training ? How did you config them ? 
Thanks in advance

**Comment by yanyanxixi** [comment]
> @yanyanxixi I think I have the same question likes you. You have used teacher-forcing in tacotron training or hifigan training ? How did you config them ?
> Thanks in advance

just do it like training step,  can get the predict_mel with the same number of real_mel,   so it can get the loss

**Comment by jik876** [comment]
I close this as there are no recent updates. Please reopen if anyone needs additional comments.

**Comment by v-nhandt21** [comment]
> > @yanyanxixi I think I have the same question likes you. You have used teacher-forcing in tacotron training or hifigan training ? How did you config them ?
> > Thanks in advance
> 
> just do it like training step, can get the predict_mel with the same number of real_mel, so it can get the loss

Hi @yanyanxixi  , I have the same problem with the miss-match, can you explain more about the step you get the fit lenght mel

- Do you change anything from the origin hifigan repo
-  How you get mel from tacotron2 ? 

Hope for your teach!

**Comment by leminhnguyen** [comment]
Hi @v-nhandt21, I've solved the problem based on the repo: https://github.com/Yeongtae/tacotron2/blob/master/GTA.py, but it has some errors need to be fixed. if you cannot figure out, you can ping me, I will find the code which has been modified by me and give you the link 😅.

**Comment by v-nhandt21** [comment]
> Hi @v-nhandt21, I've solved the problem based on the repo: https://github.com/Yeongtae/tacotron2/blob/master/GTA.py, but it has some errors need to be fixed. if you cannot figure out, you can ping me, I will find the code which has been modified by me and give you the link sweat_smile.

Wow, Is GTA really work and match size with the mel from ground-truth audio?

But I still wonder the difference between the two ways of generating mel, then fine-tune by force it learn with L1 :)) Have you compare element (mel frame) of them.
And Do we need a quality Tacotron2 model first?
Let's me try and I will ping you later for help!

**Comment by leminhnguyen** [comment]
@v-nhandt21 

> Wow, Is GTA really work and match size with the mel from ground-truth audio?
But I still wonder the difference between the two ways of generating mel, then fine-tune by force it learn with L1 :)) Have you compare element (mel frame) of them.

Yes it's really work and improves the quality of audio more better. Indeed the `gta` came from the training phase of tacotron, tacotron is an auto-regressive model so the successive mel-frame will depend on the previous, so you can only generate one mel-frame at one time. In the inference phase, the model will predict the next frame based on the frame which has been predicted, but in the training the model will use the ground truth mel-frame to predict the next one, so the size of the predicted mel can match with the ground-truth. I remember when I used `gta` to handle the miss-match problem, there were some audios still have the miss-match of 1 frame, but the quantity is small so you can freely remove those.

> And Do we need a quality Tacotron2 model first?

For my experiments you should have the good Tacotron model first, then generating mel-frame by teaching-force mode and finetuning the HifiGAN model with the generated mels and the quality of audio should have improved more better.

**Comment by v-nhandt21** [comment]
> eriments you should have the good Tacotron model first, then generating mel-frame by teaching-force mode and finetuning the HifiGAN model with the generated mels and the qualit

Yes, I also have 1 zero-frame, I guess that mel from AR model need one starting frame for init the mel generating process (I don't know my statement is correct or not), thereforce I remove the first frame of all mel from teacher forcing

![image](https://user-images.githubusercontent.com/36092323/137179876-a2aaaf9f-15ed-4d82-a3c2-74b7be34bdf1.png)

I also experiment with multiband-hifigan implementation and hope to solve this.

 anw, thank you too much @leminhnguyen , hope I can meet you in the future <3

**Comment by leminhnguyen** [comment]
@v-nhandt21 
yeah, I hope so 😁
